### PR TITLE
fix: added incremental typing to piston

### LINKED ISSUE #8: refactor: rewrite in typescript

### PR DESCRIPTION
**Pull request by Enderchief**

### CODE REVIEW

File: api/src/api/v2.js [hunk 1]
```diff
@@ -203,6 +205,7 @@ router.ws('/connect', async (ws, req) => {
 
     ws.on('message', async data => {
         try {
+            // @ts-ignore
             const msg = JSON.parse(data);
```
Comment (Endercheif) [inline api/src/api/v2.js:209]: JSON.parse takes in `string` but `data` is typed `Buffer | ArrayBuffer | Buffer[]`. Internally it invokes `toString` and in production it does work.

File: api/src/api/v2.js [hunk 2]
```diff
@@ -1,12 +1,14 @@
...
-const events = require('events');
+import { EventEmitter } from 'events';
 
-const runtime = require('../runtime');
-const { Job } = require('../job');
-const package = require('../package');
-const logger = require('logplease').create('api/v2');
+import { get_latest_runtime_matching_language_version, runtimes as _runtimes } from '../runtime';
+import Job from '../job';
+import package_ from '../package';
```
Comment (Endercheif) [inline api/src/api/v2.js:7]: `package` is a reserved word in strict mode

File: api/src/api/v2.js [hunk 3]
```diff
@@ -300,25 +303,25 @@ router.get('/runtimes', (req, res) => {
 
 router.get('/packages', async (req, res) => {
     logger.debug('Request to list packages');
-    let packages = await package.get_package_list();
+    let packages = await package_.get_package_list();
 
-    packages = packages.map(pkg => {
```
Comment (Endercheif) [inline api/src/api/v2.js:305]: both these `packages` have different types so they should be different variables

File: api/src/index.js [hunk 4]
```diff
@@ -1,72 +1,71 @@
...
+const logger = create('index', {});
 const app = express();
 expressWs(app);
 
 (async () => {
-    logger.info('Setting loglevel to', config.log_level);
-    Logger.setLogLevel(config.log_level);
+    logger.info('Setting loglevel to');
+    // @ts-ignore
+    setLogLevel(config.log_level);
```
Comment (Endercheif) [inline api/src/index.js:20]: `config.log_level` is typed `string`
`setLogLevel` is typed `"INFO" | "DEBUG" | "WARN" | "ERROR" | "NONE"`

File: api/src/index.js [hunk 5]
```diff
@@ -86,7 +85,7 @@ expressWs(app);
     logger.debug('Calling app.listen');
     const [address, port] = config.bind_address.split(':');
 
-    app.listen(port, address, () => {
+    app.listen(+port, address, () => {
```
Comment (Endercheif) [inline api/src/index.js:88]: `port` is `string` but `app.listen` takes in `number`

File: api/src/job.js [hunk 6]
```diff
@@ -323,35 +327,36 @@ class Job {
         while (processes.length > 0) {
             processes = [];
 
-            const proc_ids = fss.readdir_sync('/proc');
+            const proc_ids = readdirSync('/proc');
 
             processes = proc_ids.map(proc_id => {
-                if (isNaN(proc_id)) return -1;
+                if (isNaN(+proc_id)) return -1;
```
Comment (Endercheif) [inline api/src/job.js:333]: `isNaN` takes `number`

File: api/src/job.js [hunk 7]
```diff
@@ -323,35 +327,36 @@ class Job {
...
-                    
-                    const proc_id_int = parse_int(proc_id);
-                    
+
+                    const proc_id_int = parseInt(proc_id);
+
                     // Skip over any processes that aren't ours.
-                    if(ruid != this.uid && euid != this.uid) return -1; 
+                    // @ts-ignore: dont want to risk fixing this
+                    if (ruid != this.uid && euid != this.uid) return -1;
```
Comment (Enderchief) [inline api/src/job.js:395]: I'm scared to change this

File: api/src/runtime.js [hunk 8]
```diff
@@ -99,7 +102,8 @@ class Runtime {
 
     static load_package(package_dir) {
         let info = JSON.parse(
-            fss.read_file_sync(path.join(package_dir, 'pkg-info.json'))
+            // @ts-ignore
+            readFileSync(join(package_dir, 'pkg-info.json'))
         );
```
Comment (Endercheif) [inline api/src/runtime.js:107]: `JSON.parse` takes a string

**Comment by Enderchief** [comment]
Leaving comments for future me

**Comment by Enderchief** [comment]
:rocket:

**Pull request merged**